design/20200331.acme-self-check-failure-policy.md

```diff
@@ -0,0 +1,89 @@
+- [Proposal](#proposal)
+  - [API Changes](#api-changes)
+  - [Controller Behaviour](#controller-behaviour)
+  - [Internal API Resource Behaviour](#internal-api-resource-behaviour)
+  - [Test Plan](#test-plan)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+      - [Alpha](#alpha)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+      - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+      - [Removing a deprecated flag](#removing-a-deprecated-flag)
+  - [Version Skew Strategy](#version-skew-strategy)
+<!-- /toc -->
+
+## Summary
+
+When requesting certificates from Lets Encrypt, cert-manager performs a self-check to make sure that we have properly configured the ACME challenge response for the domain. This is to avoid redundant calls to Lets Encrypt and avoid getting rate limited. This is very good safety net. However, it would be a great addition to cert-manager if this check could be disabled or at least configured to allow failure.
+
+## Motivation
+
+In certain cluster setups, the ACME challenge self-check performed by cert-manager isn't always possible. For example, when using MetalLB and you have local DNS records that points towards the IP address assigned by MetalLB you might run into some hairpin issues, causing the traffic to fail.
+
+There is an issue describing some of the problems: [#1292](https://github.com/jetstack/cert-manager/issues/1292)
+
+### Goals
+
+* Add support for allowing the self-check to fail
+
+## Proposal
+
+We need a new policy enum to control the failure policy. It should be specified on the ACME solver level, so it works for both DNS01 and HTTP01. The two values in the initial version would be:
+* `Fail` - the self check must pass before the request is sent.
+* `Ignore` - the request will be sent even if the self check fails.
+
+### API changes
+
+To specify how ACME challenge self-checks are allowed to fail, we need to slightly modify the API.
+
+```go
+package v1
+
+// Describes how the ACME challenge self-check behaves when it fails.
+// +kubebuilder:validation:Enum=Fail;Ignore
+type ACMEFailurePolicy string
+
+const (
+	// The default failure policy. This policy causes the request to fail before it is sent to the certificate provider.
+	ACMEFailurePolicyFail ACMEFailurePolicy = "Fail"
+
+	// This policy will cause the ACME challenge self-check response to be ignored and the request will be sent to the certificate provider anyway.
+	ACMEFailurePolicyIgnore ACMEFailurePolicy = "Ignore"
+)
+```
+
+```go
+package v1alpha2
+
+type ACMEChallengeSolver struct {
+	// EXISTING FIELDS HERE
+
+	// ADDITIONAL FIELDS
+
+	// Controls how the self check behaves upon failure.
+	// +optional
+	FailurePolicy cmmeta.ACMEFailurePolicy `json:"failurePolicy,omitempty"`
+}
+```
+
+## Controller behaviour
+
+If the self-check has failed the maximum number of attempts and this policy is set to `Ignore`, perform the request the certificate provider anyway.
```

Comment:
Can you perhaps note what needs to chance in order to support this? i.e. which controllers will need modifications (I'd guess just the 'acmechallenges' controller).

Also, for the "testing" side of things, we should note:

1) changes to unit tests to cover this case
2) if possible, an end-to-end test case that ensures that a 'failing' self check (could just be one that is not properly configured) is still attempted to be 'accepted'. I think this could be done by creating a request for a domain that we *cannot* properly pass the self check for, and then waiting to ensure that the challenge is marked as `failed` (instead of `pending` which is what it'd do in the case of the policy being `Fail`)

Reply:
Notes added.